lnwallet/rpcwallet/remote_signer_client.go

```diff
@@ -0,0 +1,882 @@
+}
+
+// Stop stops the remote signer client.
+func (r *OutboundClient) Stop() error {
+	if !atomic.CompareAndSwapInt32(&r.stopped, 0, 1) {
+		return errors.New("remote signer client is already shut down")
+	}
+
+	log.Info("Remote signer client shutting down")
+
+	// Ensure that no new Run goroutines can start when we've initiated
+	// the stopping of the remote signer client.
+	r.wgMu.Lock()
+
+	if r.streamFeeder != nil {
+		r.streamFeeder.Stop()
+	}
+
+	close(r.quit)
+
+	r.wgMu.Unlock()
+
+	r.wg.Wait()
+
+	log.Debugf("Remote signer client shut down")
+
+	return nil
+}
+
+// run creates a new stream to the watch-only node, and starts processing and
+// responding to the sign requests that are sent over the stream. The function
+// will continuously run until the remote signer client is either stopped or
+// the stream errors.
+func (r *OutboundClient) run() error {
+	r.wgMu.Lock()
+
+	select {
+	case <-r.quit:
+		r.wgMu.Unlock()
+		return ErrShuttingDown
+	default:
+	}
+
+	r.wgMu.Unlock()
+
+	streamCtx, cancel := context.WithCancel(context.Background())
+
+	// Cancel the stream context whenever we return from this function.
+	defer cancel()
+
+	log.Infof("Attempting to setup the watch-only node connection")
+
+	// Try to get a new stream to the watch-only node.
+	stream, streamCleanUp, err := r.streamFeeder.GetStream(streamCtx)
+	if err != nil {
+		return err
+	}
+
+	r.stream = stream
+	defer streamCleanUp()
+
+	// Once the stream has been created, we'll need to perform the handshake
+	// process with the watch-only node, before it will start sending us
+	// requests.
+	err = r.handshake(streamCtx)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Completed setup connection to watch-only node")
+
+	// Reset the retry timeout after a successful connection.
+	r.retryTimeout = defaultRetryTimeout
+
+	return r.processSignRequests(streamCtx)
+}
+
+// handshake performs the handshake process with the watch-only node. As we are
+// the initiator of the stream, we need to send the first message over the
+// stream. The watch-only node will only proceed to sending us requests after
+// the handshake has been completed.
+func (r *OutboundClient) handshake(streamCtx context.Context) error {
+	var (
+		regSentChan = make(chan struct{})
+		regDoneChan = make(chan *walletrpc.SignCoordinatorRequest)
+		errChan     = make(chan error)
+
+		// The returnedChan is used to signal that this function has
+		// already returned, so that the goroutines don't remain blocked
+		// indefinitely when trying to send to the above channels.
+		returnedChan = make(chan struct{})
+	)
+	defer close(returnedChan)
+
+	// TODO(viktor): This could be extended to include info about the
+	// version of the remote signer in the future.
+	regType := &walletrpc.SignCoordinatorResponse_SignerRegistration{
+		SignerRegistration: "outboundSigner",
+	}
+
+	registrationMsg := &walletrpc.SignCoordinatorResponse{
+		RefRequestId:     handshakeRequestID,
+		SignResponseType: regType,
```

Comment:
After our offline discussions, I think we concluded that this could prove in terms of auth, would be that it could be make the signer be sure that it has connected to an authentic watch-only node. The watch-only node on the other hand should know that the signer is authentic as it has a valid macaroon needed to connect to the watch-only node.

Therefore I updated the content of the registration messages in the latest push, so that the Signer can pass a challenge string to the watch-only node with it's registration message. I also updated the registration response message so that the watch-only node can pass a signature in that message, which should be the signature produced when the watch-only node signs the challenge message.

These fields will not be correctly set in this PR though, as I'd like to leave that to a follow-up PR.
The addition of the fields did make the code a bit more verbose in the handshake though, but I think it's worth adding them for making it possible to add this support in the future.

Please check the latest updates and see if you agree with my changes, or if you have another suggestion of how we should future-proof this in this PR!

Alternatively, if you think it's unnecessary that the signer should be able to validate that it has connected to an authentic watch-only node, let me know! I guess given that the signer set's the RPC host it's connecting to, that could be argued, even though there still potential for spoofing or man-in-the-middling with that option.